### ISSUE TITLE
Number of sub batches

### ISSUE THREAD

**Issue by xu-ji** [vector-1127/DAC#6]
https://github.com/vector-1127/DAC/blob/2246b1dcafad4e57c0a8dda3dd4a3321836979a7/IMAGENET10/imagenet10.py#L237

Was this supposed to be `if sign>=1`? It's a little misleading right now, it looks like you are expecting 2 loops...

Also was python's `next()` what you were looking for here? It seems you can avoid the loop altogether. Unless I misunderstood.

**Comment by vector-1127** [comment]
Yes, we execute 2 loops.
This is just a hyper-parameter setting. I think "if sign>0" is also practicable.

It is not python's next(). It is an function in myImage.py, which is similar to the ImageDataGenerator in Keras.
https://keras.io/preprocessing/image/

**Comment by vector-1127** [comment]
Yes, we execute 2 loops.
This is just a hyper-parameter setting. "if sign>0" is also practicable.

It is not python's next(). It is an function in myImage.py, which is similar to the ImageDataGenerator in Keras.
https://keras.io/preprocessing/image/

**Comment by xu-ji** [comment]
I don't think you execute 2 loops because `X_batch ` is already size `batch_size `. So the iterator should stop after 1 iteration. So loop is currently stopped by the iterator, not the `sign > 1` break, because sign is never `> 1`. EDIT: this is not right but still the point stands.

It seems you use exactly keras's ImageDataGenerator, at least this is the class of `datagen`.

What I was suggesting is, instead of :
```
sign = 0 
for X_batch_i in datagen.flow(X_batch, batch_size=batch_size,shuffle=False):
  loss = model.train_on_batch([X_batch_i],[Y_, Y_batch])
  sign += 1
  if sign>1:
    break
```
I think you could just replace this with
```
loss = model.train_on_batch([next(datagen.flow(X_batch, batch_size=batch_size,shuffle=False))],
                            [Y_, Y_batch])
```

**Comment by vector-1127** [comment]
In Keras,

for X_batch_i in datagen.flow(X_batch, batch_size=batch_size,shuffle=False):
  loss = model.train_on_batch([X_batch_i],[Y_, Y_batch])

This loop will not stop, if there is not "break" to stop it.

**Comment by xu-ji** [comment]
`next(datagen.flow(X_batch, batch_size=batch_size,shuffle=False))` gets the first batch from the iterator returned by flow() and then stops.

But I see now, you actually want to allow each sub batch to be processed more than once. I believe this is nonstandard, what is the benefit? Shouldn't the learning rate be increased instead, if one pass is insufficient?

**Comment by vector-1127** [comment]
We want to learn better from a batch.

And I will check your problem, and update more standard code. Thank you for your pointing out this problems.
